### PR TITLE
Remove low feerate txes when mempool full, discourage chains

### PR DESCRIPTION
**Pull request by dgenr8**
This mempool/relay policy limits the footprint of the memory pool.  The actual policy is contained in the last two commits.  Everything else is the cherry-picks from core that are necessary.

After a transaction is added to mempool:
 - Remove expired transactions
 - If needed, remove the lowest-feerate transactions until dynamic memory usage is below the limit set by -maxmempool (default 300MB, as in https://github.com/bitcoin/bitcoin/pull/6722).

The effective feerate for a transaction is defined as

fee(transaction) / (size(transaction) + size(ancestors))

In other words, every tx is expected to pay for the size of its required unconfirmed ancestors, in addition to its own size. Beneficial effects of this rule are:
 - Unconfirmed chains are directly discouraged, in proportion to their length in bytes
 - Memory attacks involving high-fee ancestors and low-fee tx, or vice versa, are avoided
 - The effect fades away as the ancestors are confirmed

Based on changes in bitcoin core, plus additional work from @sdaftuar.

### CODE REVIEW

**Comment by zander** [comment]
no, please no.  See ML for longer answer.

**Comment by dgenr8** [comment]
This change offers these benefits for bitcoin users

 - Offers a sane transaction replacement path for users.  Wait 2 hours.<br>Wallet revert-tx (double-spend to self) and forget-tx (just consider coins available) functions are needed to assist this.

 - Maintains a user path to quick confirmation, even in the face of spam attacks. Pay a higher fee.

 - Increases the cost of mounting spam attacks.

 - Makes unconfirmed transactions more reliable by promoting mempool consistency within XT, and with core

**Comment by dgenr8** [comment]
How can the effectiveness of a mempool size-limit policy be quantified? How about: % of transactions published in blocks, that are found in local mempool at time of publication?

**Comment by ABISprotocol** [comment]
Slippery slope.  I object strongly to this.  As a wise person once said, "Doooon't doooo iiiiit."

**Comment by dgenr8** [comment]
I planning to add in Gavin's scheme to protect wallet transactions by prioritizing them.

**Comment by dgenr8** [comment]
Random eviction must be reverted for [thin blocks](https://github.com/bitcoinxt/bitcoinxt/pull/109) to go in.  Otherwise, we would be throwing away transactions, only to ask for them again when a block is received.

This PR does that reversion, and aligns the mempool policy with miner incentives.  It has been working as advertised on mainnet for me for the last several months.

**Comment by dgenr8** [comment]
Title updated to reflect the removal of aggressive expiration.  Instead of 2 hours, we now keep the much more conservative 72-hour expiration window from core.

Deploying to a mainnet node now.

**Comment by dgenr8** [comment]
Results on mainnet look good.  With mempool maxed at 300MB for over 24 hours, for the most recent blocks:

block ``000000000000000007d0e2c896d2de72b40ddfb985df184d0e47493a9527003f``
non-coinbase tx count: 1647
mempool matched: 1646
miss rate : 1/1647 = 6.07x10<sup>-4</sup>

block ``000000000000000002ef97b91702e0664a8a1a63cf3a005e109c6cc5b05bdb17``
non-coinbase tx count: 1924
mempool matched: 1918
miss rate : 6/1924 = .0031

**Comment by dagurval** [comment]
ACK
